org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/generators/AbstractXML2GrammarGenerator.java

```diff
@@ -0,0 +1,125 @@
+/*******************************************************************************
+* Copyright (c) 2020 Red Hat Inc. and others.
+* All rights reserved. This program and the accompanying materials
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lemminx.extensions.generators;
+
+import java.util.List;
+
+import org.eclipse.lemminx.commons.TextDocument;
+import org.eclipse.lemminx.dom.DOMAttr;
+import org.eclipse.lemminx.services.IXMLFormatter;
+import org.eclipse.lemminx.settings.SharedSettings;
+import org.eclipse.lemminx.utils.XMLBuilder;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.TextEdit;
+import org.w3c.dom.Attr;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.NamedNodeMap;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+
+/**
+ * Abstract class to generate a grammar (XSD, DTD, etcc) from a given XML
+ * source.
+ * 
+ * <p>
+ * The process is to build a generic {@link GrammarInfo} instance from the XML
+ * source and each implementation uses this grammar information to generate the
+ * proper grammar.
+ * </p>
+ * 
+ * @param <T> the grammar generator settings.
+ */
+public abstract class AbstractXML2GrammarGenerator<T extends FileContentGeneratorSettings>

```

Comment:
I think the generator should be decoupled from lsp4j

Reply:
fixed, now the generator returns a String instead of TextEdit. I rename IXMLFormatter to IXMLFullFomratter to return a String an ddon't use Range.